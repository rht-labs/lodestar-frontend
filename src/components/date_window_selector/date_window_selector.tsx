import {
  ToggleGroup,
  ToggleGroupItem,
  ToggleGroupVariant,
} from '@patternfly/react-core';
import {
  addQuarters,
  eachQuarterOfInterval,
  endOfQuarter,
  getQuarter,
  startOfQuarter,
  sub,
  subQuarters,
  subYears,
} from 'date-fns';
import React, { useState } from 'react';

export interface DateRange {
  startDate: Date;
  endDate: Date;
}

export interface DateWindowSelectorProps {
  onSelectWindow?(dates?: DateRange): void;
}

enum DateWindows {
  OneWeek,
  OneMonth,
  ThreeMonths,
  SixMonths,
  OneYear,
  AllTime,
  Q1,
  Q2,
  Q3,
  Q4,
}

const lastDayOfLastQuarter = endOfQuarter(subQuarters(new Date(), 1));
const quarters = eachQuarterOfInterval({
  end: lastDayOfLastQuarter,
  start: startOfQuarter(addQuarters(subYears(lastDayOfLastQuarter, 1), 1)),
}).reverse();

export const DateWindowSelector = ({
  onSelectWindow = () => {},
}: DateWindowSelectorProps) => {
  const [selectedRange, setSelectedRange] = useState<DateWindows>(
    DateWindows.AllTime
  );
  const handleClick = (dateWindow: DateWindows) => () => {
    setSelectedRange(dateWindow);
    switch (dateWindow) {
      case DateWindows.OneWeek:
        onSelectWindow({
          startDate: sub(new Date(Date.now()), { weeks: 1 }),
          endDate: new Date(Date.now()),
        });
        break;
      case DateWindows.OneMonth:
        onSelectWindow({
          startDate: sub(new Date(Date.now()), { months: 1 }),
          endDate: new Date(Date.now()),
        });
        break;
      case DateWindows.ThreeMonths:
        onSelectWindow({
          startDate: sub(new Date(Date.now()), { months: 3 }),
          endDate: new Date(Date.now()),
        });
        break;
      case DateWindows.SixMonths:
        onSelectWindow({
          startDate: sub(new Date(Date.now()), { months: 6 }),
          endDate: new Date(Date.now()),
        });
        break;
      case DateWindows.OneYear:
        onSelectWindow({
          startDate: sub(new Date(Date.now()), { years: 1 }),
          endDate: new Date(Date.now()),
        });
        break;
      case DateWindows.AllTime:
        onSelectWindow();
        break;
      default:
        const quarterSelection = quarters.find(
          q => `Q${getQuarter(q)}` === DateWindows[dateWindow]
        );
        if (!!quarterSelection) {
          onSelectWindow({
            startDate: startOfQuarter(quarterSelection),
            endDate: endOfQuarter(quarterSelection),
          });
        }
    }
  };
  return (
    <div>
      <ToggleGroup variant={ToggleGroupVariant.light}>
        <ToggleGroupItem
          text="1w"
          key={'1w'}
          data-testid="1w-window"
          onChange={handleClick(DateWindows.OneWeek)}
          isSelected={selectedRange === DateWindows.OneWeek}
        ></ToggleGroupItem>
        <ToggleGroupItem
          text="1m"
          key={'1m'}
          data-testid="1m-window"
          onChange={handleClick(DateWindows.OneMonth)}
          isSelected={selectedRange === DateWindows.OneMonth}
        ></ToggleGroupItem>
        <ToggleGroupItem
          text="3m"
          key={'3m'}
          data-testid="3m-window"
          onChange={handleClick(DateWindows.ThreeMonths)}
          isSelected={selectedRange === DateWindows.ThreeMonths}
        ></ToggleGroupItem>
        <ToggleGroupItem
          text="6m"
          key={'6m'}
          data-testid="6m-window"
          onChange={handleClick(DateWindows.SixMonths)}
          isSelected={selectedRange === DateWindows.SixMonths}
        ></ToggleGroupItem>
        <ToggleGroupItem
          text="1y"
          key={'1y'}
          data-testid="1y-window"
          onChange={handleClick(DateWindows.OneYear)}
          isSelected={selectedRange === DateWindows.OneYear}
        ></ToggleGroupItem>
        {quarters.map(q => {
          const quarter = getQuarter(q);
          return (
            <ToggleGroupItem
              text={`Q${quarter}'${q.getFullYear() - 2000}`} //"Q1'21"
              key={'q' + quarter.toString()}
              data-testid={'q' + quarter.toString()}
              onChange={handleClick(DateWindows['Q' + quarter.toString()])}
              isSelected={
                selectedRange === DateWindows['Q' + quarter.toString()]
              }
            ></ToggleGroupItem>
          );
        })}
        <ToggleGroupItem
          text="All Time"
          key={'alltime'}
          data-testid="alltime-window"
          onChange={handleClick(DateWindows.AllTime)}
          isSelected={selectedRange === DateWindows.AllTime}
        ></ToggleGroupItem>
      </ToggleGroup>
    </div>
  );
};
