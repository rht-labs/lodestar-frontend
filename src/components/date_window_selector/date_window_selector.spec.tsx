import { fireEvent, render } from '@testing-library/react';
import { sub } from 'date-fns';
import React from 'react';
import { DateRange, DateWindowSelector } from './date_window_selector';

describe('Date window selector', () => {
  test('shows 1w, 1m, 3m, 6m, 1y, All time selectors', () => {
    const component = render(<DateWindowSelector />);
    expect(component.getByTestId('1w-window')).toBeDefined();
    expect(component.getByTestId('1m-window')).toBeDefined();
    expect(component.getByTestId('3m-window')).toBeDefined();
    expect(component.getByTestId('6m-window')).toBeDefined();
    expect(component.getByTestId('1y-window')).toBeDefined();
    expect(component.getByTestId('alltime-window')).toBeDefined();
  });
  test('when clicking a date option, a callback is invoked with a min and max date', () => {
    const dateToCallbackMaps = {
      '1w': (date?: DateRange) => {
        const currentDate = new Date(Date.now());
        expect(date).toBeDefined();
        expect(date.startDate.getDate()).toEqual(
          sub(currentDate, { weeks: 1 }).getDate()
        );
      },
      '1m': (date?: DateRange) => {
        const currentDate = new Date(Date.now());
        expect(date).toBeDefined();
        expect(date.startDate.getDate()).toEqual(
          sub(currentDate, { months: 1 }).getDate()
        );
      },
      '3m': (date?: DateRange) => {
        const currentDate = new Date(Date.now());
        expect(date).toBeDefined();
        expect(date.startDate.getDate()).toEqual(
          sub(currentDate, { months: 3 }).getDate()
        );
      },
      '6m': (date?: DateRange) => {
        const currentDate = new Date(Date.now());
        expect(date).toBeDefined();
        expect(date.startDate.getDate()).toEqual(
          sub(currentDate, { months: 6 }).getDate()
        );
      },
      '1y': (date?: DateRange) => {
        const currentDate = new Date(Date.now());
        expect(date).toBeDefined();
        expect(date.startDate.getDate()).toEqual(
          sub(currentDate, { years: 1 }).getDate()
        );
      },
      'All Time': (date?: DateRange) => {
        expect(date).toBeUndefined();
      },
    };
    let handleSelectWindow = jest.fn();
    const component = render(
      <DateWindowSelector onSelectWindow={handleSelectWindow} />
    );
    for (let key in dateToCallbackMaps) {
      handleSelectWindow = jest
        .fn()
        .mockImplementation(dateToCallbackMaps[key]);
      component.rerender(
        <DateWindowSelector onSelectWindow={handleSelectWindow} />
      );
      fireEvent.click(component.getByText(key));
      expect(handleSelectWindow).toHaveBeenCalled();
    }
  });
  test('Clicking a date window selects that window, and that window should be highlighted', async () => {
    const component = render(<DateWindowSelector onSelectWindow={() => {}} />);
    fireEvent.click(component.getByText('1w'));
    expect(
      component
        .getByTestId('1w-window')
        .getElementsByTagName('button')[0]
        .className.split(' ')
    ).toContain('pf-m-selected');
  });
});
