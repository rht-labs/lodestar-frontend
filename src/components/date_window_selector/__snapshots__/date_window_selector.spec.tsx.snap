// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date window selector matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="pf-c-toggle-group"
          role="group"
        >
          <div
            class="pf-c-toggle-group__item"
            data-testid="1w-window"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                1w
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="1m-window"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                1m
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="3m-window"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                3m
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="6m-window"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                6m
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="1y-window"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                1y
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="q1"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                Q1'21
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="q4"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                Q4'20
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="q3"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                Q3'20
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="q2"
          >
            <button
              aria-pressed="false"
              class="pf-c-toggle-group__button pf-m-light"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                Q2'20
              </span>
            </button>
          </div>
          <div
            class="pf-c-divider pf-m-vertical"
            role="separator"
          />
          <div
            class="pf-c-toggle-group__item"
            data-testid="alltime-window"
          >
            <button
              aria-pressed="true"
              class="pf-c-toggle-group__button pf-m-light pf-m-selected"
              type="button"
            >
              <span
                class="pf-c-toggle-group__text"
              >
                All Time
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="pf-c-toggle-group"
        role="group"
      >
        <div
          class="pf-c-toggle-group__item"
          data-testid="1w-window"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              1w
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="1m-window"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              1m
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="3m-window"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              3m
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="6m-window"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              6m
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="1y-window"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              1y
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="q1"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              Q1'21
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="q4"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              Q4'20
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="q3"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              Q3'20
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="q2"
        >
          <button
            aria-pressed="false"
            class="pf-c-toggle-group__button pf-m-light"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              Q2'20
            </span>
          </button>
        </div>
        <div
          class="pf-c-divider pf-m-vertical"
          role="separator"
        />
        <div
          class="pf-c-toggle-group__item"
          data-testid="alltime-window"
        >
          <button
            aria-pressed="true"
            class="pf-c-toggle-group__button pf-m-light pf-m-selected"
            type="button"
          >
            <span
              class="pf-c-toggle-group__text"
            >
              All Time
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
